### ISSUE TITLE
Failing to bring up resources on apply

### ISSUE THREAD

**Issue by sps-software** [cn-terraform/terraform-aws-ecs-fargate-service#13]
I'm getting this error when I apply:

```
Error: Invalid count argument

 on .terraform/modules/ecs-fargate-service/terraform-aws-ecs-fargate-service-2.0.3/main.tf line 113, in resource "aws_security_group_rule" "ingress_through_http":
 113:   count                    = length(data.aws_lb_target_group.lb_http_target_groups)

The "count" value depends on resource attributes that cannot be determined
until apply, so Terraform cannot predict how many instances will be created.
To work around this, use the -target argument to first apply only the
resources that the count depends on.
```
I can probably get around this by running apply with each lb target group data source as the target, but I think it might help to either add that to the readme or update it... Unless I'm incorrectly using the module?

**Comment by rredpoppy** [comment]
any solution to this? anyway I can help?

**Comment by jnonino** [comment]
Hi, I have to review the chain of resources to validate dependencies and see where the issue is. 
Lately I've been busy and do not have the time to work on these issues. But as you can see in this module and the [other modules](https://github.com/cn-terraform) every collaboration is welcome. Some people have created PRs solving issues or improving the modules and I try to review them and release a new module with the changes as soon as possible.
I definitely need to find some time to fix this and other issues in these modules, but if you know where the problem is and if you have time to fill a PR, your contribution is always welcome!!!

**Comment by rredpoppy** [comment]
problem seems to be in the architecture of the whole thing. terraform has introduced [some issues with count](https://github.com/hashicorp/terraform/issues/25851). It might also be, as one use mentions it, that there is an issue with the key of the resource. I don't know how to solve this, I would have forked it already.

**Comment by jnonino** [comment]
Yes, that was what I imagined. The issue seems to be with count, I will check using for_each or something like that.

**Comment by rredpoppy** [comment]
trying it now with foreach on my fork, saw a clue in the mentioned terraform issue

**Comment by jnonino** [comment]
Do you have time to test it using something like this?

```
module "ecs-fargate-service" {
  source  = "git@github.com:cn-terraform/terraform-aws-ecs-fargate-service.git?ref=issue13/failing-to-bring-up-resources-on-apply"
  # insert the 14 required variables here
}
```

**Comment by rredpoppy** [comment]
testing now, it fails because of list being null if https is disabled:
```
 on .terraform/modules/kopplr_graphql_service/main.tf line 124, in resource "aws_security_group_rule" "ingress_through_https":
 124:   for_each                    = toset(data.aws_lb_target_group.lb_https_target_groups)
    |----------------
    | data.aws_lb_target_group.lb_https_target_groups is object with no attribute
```
so maybe some additional null testing with an empty set default?

**Comment by jnonino** [comment]
Done, how about now??

**Comment by rredpoppy** [comment]
```
Error: Invalid function argument

  on .terraform/modules/kopplr_graphql_service/main.tf line 114, in resource "aws_security_group_rule" "ingress_through_http":
 114:   for_each                 = toset(data.aws_lb_target_group.lb_http_target_groups)
    |----------------
    | data.aws_lb_target_group.lb_http_target_groups is object with 1 attribute "arn:aws:elasticloadbalancing:eu-central-1:256227198371:targetgroup/kopplr-alb-lb-http-tg-0/1243b56dfc393749"

Invalid value for "v" parameter: cannot convert object to set of any single
type.


Error: Invalid function argument

  on .terraform/modules/kopplr_graphql_service/main.tf line 124, in resource "aws_security_group_rule" "ingress_through_https":
 124:   for_each                 = toset(data.aws_lb_target_group.lb_https_target_groups)
    |----------------
    | data.aws_lb_target_group.lb_https_target_groups is object with no attributes

Invalid value for "v" parameter: cannot convert object to set of any single
type.
```

**Comment by jnonino** [comment]
Thanks for helping, what about now?

**Comment by rredpoppy** [comment]
```
Error: Invalid count argument

  on .terraform/modules/kopplr_graphql_service/main.tf line 114, in resource "aws_security_group_rule" "ingress_through_http":
 114:   count                    = length(data.aws_lb_target_group.lb_http_target_groups)

The "count" value depends on resource attributes that cannot be determined
until apply, so Terraform cannot predict how many instances will be created.
To work around this, use the -target argument to first apply only the
resources that the count depends on.


Error: Invalid count argument

  on .terraform/modules/kopplr_graphql_service/main.tf line 124, in resource "aws_security_group_rule" "ingress_through_https":
 124:   count                    = length(data.aws_lb_target_group.lb_https_target_groups)

The "count" value depends on resource attributes that cannot be determined
until apply, so Terraform cannot predict how many instances will be created.
To work around this, use the -target argument to first apply only the
resources that the count depends on.
```

**Comment by rredpoppy** [comment]
```
Error: Invalid "each" attribute

  on .terraform/modules/kopplr_graphql_service/main.tf line 128, in resource "aws_security_group_rule" "ingress_through_https":
 128:   to_port                  = each.port
```

**Comment by jnonino** [comment]
Ok, I pushed a change, one more chance!!!

**Comment by rredpoppy** [comment]
```
Error: Unsupported attribute

  on .terraform/modules/kopplr_graphql_service/main.tf line 114, in resource "aws_security_group_rule" "ingress_through_http":
 114:   for_each                 = data.aws_lb_target_group.lb_http_target_groups != null ? toset(data.aws_lb_target_group.lb_http_target_groups.*.port) : toset([])

This object does not have an attribute named "port".


Error: Unsupported attribute

  on .terraform/modules/kopplr_graphql_service/main.tf line 124, in resource "aws_security_group_rule" "ingress_through_https":
 124:   for_each                 = data.aws_lb_target_group.lb_https_target_groups != null ? toset(data.aws_lb_target_group.lb_https_target_groups.*.port) : toset([])

This object does not have an attribute named "port".
```

it's actually a bit more involved than that, [I am trying to figure it out mysef](https://github.com/WhiteCityCode/terraform-aws-ecs-fargate-service/blob/master/main.tf). for me it's complaining about being a set of objects and expecting a set of strings

**Comment by rredpoppy** [comment]
I think the resolution would be to actually inject the target groups instead of getting them via data resources based on the arns

**Comment by jnonino** [comment]
I made a change, can you confirm it works now?

**Comment by rredpoppy** [comment]
at least for my case (https disabled) this works. thank you

**Comment by jnonino** [comment]
Great, thanks for your help!!!
Released in version [2.0.9](https://registry.terraform.io/modules/cn-terraform/ecs-fargate-service/aws/2.0.9)
